jetty-client/src/main/java/org/eclipse/jetty/client/util/AbstractRequestContent.java

```diff
@@ -0,0 +1,250 @@
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.client.util;
+
+import java.io.EOFException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractRequestContent implements Request.Content
+{
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractRequestContent.class);
+
+    private final AutoLock lock = new AutoLock();
+    private final String contentType;
+    private Subscription subscription;
+    private Throwable failure;
+
+    protected AbstractRequestContent(String contentType)
+    {
+        this.contentType = contentType;
+    }
+
+    @Override
+    public String getContentType()
+    {
+        return contentType;
+    }
+
+    @Override
+    public Subscription subscribe(Consumer consumer, boolean emitInitialContent)
+    {
+        Subscription oldSubscription;
+        Subscription newSubscription;
+        try (AutoLock ignored = lock.lock())
+        {
+            if (subscription != null && !isReproducible())
+                throw new IllegalStateException("Multiple subscriptions not supported on " + this);
+            oldSubscription = subscription;
+            newSubscription = subscription = newSubscription(consumer, emitInitialContent, failure);
+        }
+        if (oldSubscription != null)
+            oldSubscription.fail(new EOFException("Content replay"));
+        if (LOG.isDebugEnabled())
+            LOG.debug("Content subscription for {}: {}", this, consumer);
+        return newSubscription;
+    }
+
+    protected abstract Subscription newSubscription(Consumer consumer, boolean emitInitialContent, Throwable failure);
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        Subscription subscription = null;
+        try (AutoLock ignored = lock.lock())
+        {
+            if (this.failure == null)
+            {
+                this.failure = failure;
+                subscription = this.subscription;
+            }
+        }
+        if (subscription != null)
+            subscription.fail(failure);
+    }
+
+    public abstract class AbstractSubscription implements Subscription
+    {
+        private final Consumer consumer;
+        private final boolean emitInitialContent;
+        private Throwable failure;
+        private int demand;
+        private boolean stalled;
+        private boolean committed;
```

Comment:
needs javadoc (and perhaps a better names?) for what fields mean, specially `emitInitialContent` and `committed`.  A state diagram or table might help!

Reply:
Will do.

Reply:
`emitInitialContent` ????

Reply:
@sbordet I would still like to see javadoc on the what/why/how of `emitInitialContent`
Maybe a state diagram/table wouldn't hurt either.